azurelinuxagent/ga/exthandlers.py

```diff
@@ -276,6 +288,73 @@ def run(self):
                       message=detailed_msg)
             return
 
+    def get_artifacts_profile(self):
+        try:
+            self.artifacts_profile = self.protocol.get_artifacts_profile()
+        except Exception as e:
+            msg = u"Exception retrieving VmArtifactsProfile blob: {0}".format(ustr(e))
+            detailed_msg = '{0} {1}'.format(msg, traceback.extract_tb(get_traceback(e)))
+            logger.warn(msg)
+
+            add_event(AGENT_NAME,
+                      version=CURRENT_VERSION,
+                      op=WALAEventOperation.ExtensionProcessing,
+                      is_success=False,
+                      message=detailed_msg)
+
+    def process_goal_state(self, instantiation, sequence_number):
+        fabric_goal_state_changed, fast_track_changed = self.determine_what_changed(instantiation, sequence_number)
+        if fabric_goal_state_changed:
+            self.protocol.set_fast_track(False)
+
+        # Fabric goal state takes precedence over VMArtifactsProfile blob goal state
+        # So, process the Fabric goal state first
+        self.handle_ext_handlers(fabric_goal_state_changed)
```

Comment:
Why not call `handle_ext_handlers` only when things change? The advantage would be to save the compute cycles as almost nothing is being done when it is False.

```
if fabric_goal_state_changed:
    self.handle_ext_handlers(fabric_goal_state_changed)
```

Reply:
I would love to do this, but I'm concerned about modifying the existing behavior. Currently it isn't until handle_ext_handler that we determine whether a goal state occurred. Since this code is for the fabric goal state, I didn't want to change that behavior.